### DISCUSSION TITLE
Protect created_at and updated_at fields

### DISCUSSION THREAD

**Discussion by c00kiemon5ter** [supabase/supabase#7541] [Questions] [answered]
I bumped into [this article](https://dev.to/jdgamble555/supabase-date-protection-on-postgresql-1n91) discussing that by default an API call could override the `created_at` and `updated_at` fields on a table. At the end of the article there's an update with the suggested solution on how to protect the table data against such requests.

I wanted to raise this as a potential issue and get the core team's view on it. Should such protection be applied by default? Should this be available in the form of a shortcut to be applied on a table?

PS: I searched into the issues and the discussions on this repo but didn't find anything related. Sorry, if I missed a relevant thread, and please redirect me there if such a thread exists.

**Comment by burggraf** [comment]
You only need to write triggers to protect `created_at` and to update `updated_at` automatically if:
1.  you need this added layer of server-level (postgres-level) security
2. you're fine with the slight overhead generated by adding these triggers to your table

So, my personal opinion -- this is how it should be.  If you need to protect these fields, you can do it with triggers.  If these fields are not critical to your application, then I wouldn't bother.

**Comment by silentworks** [comment, marked as answer]
Thanks for this suggestion @c00kiemon5ter. I've raised this with the team and we will be looking into adding it to the Scripts you see when you open the SQL Editor inside of the Supabase Dashboard.

**Reply by c00kiemon5ter** [reply]
sounds great! thank you!

**Reply by NanoCode012** [reply]
This sounds great. For me, I am using `REVOKE` and `GRANT` permissions on each table to control these settings (also other columns like `id`..), basically Column Level security https://github.com/supabase/supabase/discussions/3158 which led me to https://github.com/supabase/supabase/discussions/7428#discussion-4172777. However, it's quite messy and confusing. Perhaps, the script could accept a `column` or `column array` for fields that should not be changed?

On the other hand, I am not sure what is the performance difference between using triggers and grants, but it is to my understanding that granting permissions would be better on performance since they are run first https://github.com/supabase/supabase/discussions/3158#discussioncomment-1326278 .

This may be out of the scope of this discussion, but could I suggest UI for fine tuning `GRANTS` permission? For example, for Table A, we can select (checkbox) GRANT permission for READ,UPDATE,INSERT,DELETE for each role (or at least authenticated role) for each column?
